tracer/src/Datadog.Trace/TraceContext.cs

```diff
@@ -102,6 +102,12 @@ public void CloseSpan(Span span)
 
             lock (this)
             {
+                // TODO is this the best spot to sample for single span?
```

Comment:
Maybe not because the spans's service name and operation name can still change even after the span is closed. A user can keep a reference to the span and keep making change up until it is serialized and sent to the trace agent.

If we do this during serialization (the latest possible moment), we can also inject the tags directly into the MessagePack binary instead of adding them to the span's dictionary, possibly saving some bytes in heap allocations. Just an idea...

Reply:
> If we make this span sampling decision this during serialization

Won't it be too complicated to do that with stats computation? Also, the current sampler has the same issue (service and operation name can change). So I would do everything in `TraceContext` for the sake of simplicity.

> // TODO is this the best spot to sample for single span?

Is it mandatory to do that within the lock?

Reply:
Yes, after some discussion with Zach (on slack?) I think we determined to keep it simple for now.

Reply:
> Is it mandatory to do that within the lock?

I don't think so looking at it again, I moved it out in 491b35d4fe0761df20576ae32c921435a5551733

Reply:
> Won't it be too complicated to do that with stats computation? Also, the current sampler has the same issue (service and operation name can change). So I would do everything in TraceContext for the sake of simplicity.

I'll plan on leaving all the sampling decision stuff in `TraceContext` for the time being